tests/core/abi-utils/test_abi_utils.py

```diff
@@ -173,3 +694,440 @@ def test_event_signature_to_log_topic(event_signature, expected):
+                }
+            ),
+            "(address,uint256,bytes)[3]",
+        ),
+        (
+            (
+                {
+                    "type": "uint256",
+                }
+            ),
+            "uint256",
+        ),
+        (
+            "uint256",
+            "uint256",
+        ),
+    ),
+)
+def test_collapse_if_tuple(abi_component, expected):
+    assert collapse_if_tuple(abi_component) == expected
+
+
+@pytest.mark.parametrize(
+    "abi_component",
+    (
+        (
+            {
+                "type": False,
+            }
+        ),
+    ),
+)
+def test_collapse_if_tuple_with_errors(abi_component):
+    with pytest.raises(TypeError):
+        collapse_if_tuple(abi_component)
+
+
+@pytest.mark.parametrize(
+    "fn_abi,args,kwargs,expected",
+    (
+        (
+            FN_ABI_C,
+            (
+                "0x1234567890123456789012345678901234567890",
+                b"bar",
+                "0x1234567890123456789012345678901234567890",
+            ),
+            {},
+            (
+                "0x1234567890123456789012345678901234567890",
+                b"bar",
+                "0x1234567890123456789012345678901234567890",
+            ),
+        ),
+        (
+            FN_ABI_C,
+            [],
+            {
+                "a": "0x1234567890123456789012345678901234567890",
+                "b": b"bar",
+                "c": "0x1234567890123456789012345678901234567890",
+            },
+            (
+                "0x1234567890123456789012345678901234567890",
+                b"bar",
+                "0x1234567890123456789012345678901234567890",
+            ),
+        ),
+    ),
+)
+def test_get_normalized_abi_inputs(fn_abi, args, kwargs, expected):
+    inputs = get_normalized_abi_inputs(fn_abi, args, kwargs)
+
+    assert inputs == expected
+
+
+def test_get_normalized_abi_inputs_returns_args_without_kwargs():
+    args = [1, "0x1234567890123456789012345678901234567890", b"bar"]
+    kwargs = {}
+    inputs = get_normalized_abi_inputs(FN_ABI_C, args, kwargs)
+
+    assert inputs == args
+
+
+def test_error_when_invalid_args_kwargs_combo_provided():
+    with pytest.raises(TypeError, match="Incorrect argument count."):
+        get_normalized_abi_inputs(
+            FN_ABI_E,
+            (
+                1,
+                2,
+            ),
+            {"a": 1, "b": 2},
+        )
+
+
+def test_kwargs_is_disallowed_when_merging_with_unnamed_inputs():
+    with pytest.raises(TypeError):
+        get_normalized_abi_inputs(FN_ABI_E, tuple(), {"x": 1, "y": 2})
+
+
+def test_args_works_when_merging_with_unnamed_inputs():
+    actual = get_normalized_abi_inputs(FN_ABI_E, (1, 2), {})
+    assert actual == (1, 2)
+
+
+def test_args_allowed_when_duplicate_named_inputs():
+    actual = get_normalized_abi_inputs(FN_ABI_DUPLICATE_NAMES, (1, 2, 3), {})
+    assert actual == (1, 2, 3)
+
+
+def test_kwargs_not_allowed_for_duplicate_input_names():
```

Comment:
nit: If it makes sense to, I think these can be easier to read as parametrized. If the goal was to highlight each test case separately, we can pass `ids=['test_case_1', 'test_case_2, ...]` into `@pytest.parametrize` to highlight each test case. 

```py
@pytest.mark.parametrize(
    "abi,args,kwargs,exception,msg",
    (
        (FN_ABI_DUPLICATE_NAMES, (1,), {"a": 2, "b": 3}, TypeError, "..."),
        ({"type": "fallback", "inputs": {"a": 1}}, (1,), {}, ValueError, "fallback"),
        ({"type": "receive", "inputs": {"a": 1}}, (1,), {}, ValueError, "receive"),
        ...,
    ),
    ids=[
         "kwargs_not_allowed_for_duplicate_input_names",
         "does_not_allow_fallback_types",
         "does_not_allow_receive_types",
         ...,
    ],
)
def test_get_normalized_abi_inputs_negative_cases(abi, args, kwargs, exception, msg):
    with pytest.raises(exception, match=re.escape(msg)):
        get_normalized_abi_inputs(abi, args, kwargs)
```